corehq/apps/case_search/xpath_functions/ancestor_functions.py

```diff
@@ -144,11 +144,19 @@ def _get_case_ids_from_ast_filter(context, filter_node):
         es_filter = build_filter_from_ast(filter_node, context)
 
         es_query = CaseSearchES().domain(context.domain).filter(es_filter)
-        if es_query.count() > MAX_RELATED_CASES:
-            new_query = serialize(filter_node)
-            raise TooManyRelatedCasesError(
-                gettext("The related case lookup you are trying to perform would return too many cases"),
-                new_query
-            )
```

Comment:
Not sure what would be considered "too large" but this function can make `.scroll_ids()` return a large dataset since what it would match would be for example, all cases with case type "unit" with color property='red'. 

Do either of you know if the 500k limit is to avoid running out memory (is this related to stack/heap memory?)

For context, since I wrote this function - I put the check here under the assumption that on average, a parent will have >= 1 child, there will be more children than parents. But it seems like the limit can be hit on either the parent or child side. 

Maybe there's a way to not evaluate the query until the end by creating a chained filter instead? Not sure if that's possible

Reply:
Yea so to reiterate: this check was intended to limit the returned number of cases to the user (hence the use of MAX_RELATED_CASES) but due to some oversight was placed in the middle of this in-memory JOIN instead of at the end, resulting in the actual limit varying from domain to domain (and being way less than the promised 500k). I'm not sure if whoever wrote this intended it but I guess it also ended acting as a safeguard against the scenario that @esoergel  described. SO I think this means that we'll have to drastically reduce the promised number of ancestor cases returned (maybe more like 50k) since the alternative seems to be migrating child case info into parent cases so this join doesn't need to happen. or somehow re-writing this whole in-memory join operation though I assume someone had attempted it before settling on this method.

Reply:
I'm pretty confident that limiting the size of the join was the intended purpose of this check.  `CASE_SEARCH_MAX_RESULTS` (500) is what limits the number returned to the user for case search, while I don't think the CLE is intended to have a limit (it uses `scroll_ids_to_disk_and_iter_docs` to pull very large data sets).

This limit was introduced in https://github.com/dimagi/commcare-hq/pull/22215, which split off from https://github.com/dimagi/commcare-hq/pull/22192, which says:

> Removes the ability to perform filtering based on related case lookups (e.g. age = 'youngun' and parent/age = 'elder')
> This doesn't work for very large caseloads (1M+), which makes this feature not useful for the one solutions domain that uses this, so we're removing the functionality entirely.

You wrote:
> I think this means that we'll have to drastically reduce the promised number of ancestor cases returned

Is this something we communicate to the user?  Where does that happen?  (it's a hard concept to explain to end users)

Reply:
sorry for the delay, but my bad, I forgot that the return limit isn’t something we tell users anywhere. Not sure if it matters, but CLE does have a limit, at least for [subcase function queries](https://github.com/dimagi/commcare-hq/blob/900eeb1453b14e9c3908416f24442cd97c72eced/corehq/apps/case_search/xpath_functions/subcase_functions.py#L107), which is why I was trying to make the ancestor case search behave similarly. But based on that comment about it not working for very large caseloads, I guess there really isn’t a way to safely implement a precise limit with the current process. And even if the allowed maximum was like a million, limiting the middle step means it won't guarantee it'll capture all the possible parent case_ids for the next step. This change was mostly so we'd have a target number to reach for during performance tests but if we're comfortable having a variable "maximum" returned cases for ancestor function queries I'm fine with not going through with these changes.

Reply:
On second thought, it might be good to have a maximum limit in place after both steps, in the off chance the second, child case searching step, does return more than 500k cases.

Reply:
I agree with all that - I do think it's necessary to have a maximum intermediate query size.  I don't think that size really tells us much about the size of the second query.  Projects may have one or two child cases per parent, or a thousand per parent.  Limiting the size of the final query is also probably necessary, though the limit depends on the use-case.  For case search, it's 500, which is quite low, while the CLE probably needs multiple limits depending on how it's being accessed.  Ideally that'd probably be server-side paginated, so the max limit for a single query would be like 100, or we could use client-size pagination and an arbitrarily large cap on results returned.  If the CLE is being exported to excel, we may use no limit at all, or perhaps a limit based on what size of excel file is considered usable.